core/shared/src/main/scala/fs2/retry.scala

```diff
@@ -0,0 +1,57 @@
+package fs2
+
+import cats.effect.Async
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+
+object retry {
+  /**
+   * Retries `fa` on failure, returning a single element
+   * stream with the result of `fa` as soon as it succeeds.
+   *
+   * @param delay Duration of delay before the first retry
+   *
+   * @param nextDelay Applied to the previous delay to compute the
+   *                  next, e.g. to implement exponential backoff
+   *
+   * @param maxRetries Number of attempts before failing with the
+   *                   latest error, if `fa` never succeeds
+   *
+   * @param retriable Function to determine whether a failure is
+   *                  retriable or not, defaults to retry every
+   *                  `NonFatal`. A failed stream is immediately
+   *                  returned when a non-retriable failure is
+   *                  encountered
+   */
+  def retry[F[_], A](fa: F[A],
+                     delay: FiniteDuration,
```

Comment:
Currently I use the following implementation exactly because I need safe resource allocation and standard combinators (especially `interruptWhen`). Also I need to do things like "log every 3rd failure".
```scala
    /**
      * Retries `fa` if it fails.
      * Returns a single (the first) value produced by `fa` on a successful
      * attempt or fails with the last error if `nextDelay` results in `None`.
      *
      * @param initialDelay delay after the first attempt
      * @param nextDelay function used to compute consequent delays
      *                  `None` terminates attempts
      * @param onFailedAttempt can be used to handle (but not recover) attempt failure
      * @return the first value produced by `fa`
      */
    def retryWithDelay[F[_]: Async, A](
      fa: Stream[F, A],
      initialDelay: FiniteDuration,
      nextDelay: (Int, FiniteDuration) => Option[FiniteDuration],
      onFailedAttempt: (Int, Throwable) => Stream[F, Nothing]
    )(implicit sch: Scheduler): Stream[F, A] =
      fa.attempt.repeat.mapAccumulate((1, initialDelay)) {
        case (t@(attempts, delay), Left(ex)) =>
          nextDelay(attempts + 1, delay) match {
            case Some(ndelay) =>
              val s = onFailedAttempt(attempts, ex) ++ fs2.time.sleep_[F](delay)
              ((attempts + 1, ndelay), s)
            case None =>
              (t, Stream.fail(ex))
          }
        case (t, Right(a)) => (t, Stream.emit(a))
      }.flatMap(_._2).head
```